kernel/src/syscall/poll.rs

```diff
@@ -56,24 +59,27 @@ pub fn sys_poll(fds: Vaddr, nfds: u64, timeout: i32, ctx: &Context) -> Result<Sy
 }
 
 pub fn do_poll(poll_fds: &[PollFd], timeout: Option<Duration>, ctx: &Context) -> Result<usize> {
-    let (result, files) = hold_files(poll_fds, ctx);
-    match result {
-        FileResult::AllValid => (),
-        FileResult::SomeInvalid => {
-            return Ok(count_all_events(poll_fds, &files));
+    loop {
+        let file_table = ctx.process.file_table().lock();
+        let (result, files) = hold_files(&file_table, poll_fds);
+        match result {
+            FileResult::AllValid => (),
+            FileResult::SomeInvalid => {
+                return Ok(count_all_events(poll_fds, &files));
+            }
         }
-    }
 
-    let poller = match register_poller(poll_fds, files.as_ref()) {
-        PollerResult::AllRegistered(poller) => poller,
-        PollerResult::EventFoundAt(index) => {
-            let next = index + 1;
-            let remaining_events = count_all_events(&poll_fds[next..], &files[next..]);
-            return Ok(1 + remaining_events);
-        }
-    };
+        let poller = match register_poller(poll_fds, &files) {
```

Comment:
Thanks for your advice! 

> This forces File::poll to run in atomic mode (since we've held the file_table().lock()).

> - In our current codebase, this is wrong because some file types try to grab a mutex before touching the pollee > > > (e.g. the UNIX sockets).
> - In the future, if at least part of https://github.com/asterinas/asterinas/issues/1356 can be accepted, things will get > more complex, as calculating I/O events in FileLike::poll will just require us to grab more mutexes.

Yes, this is a real problem. I just don't take into account that `Filelike::poll` can only lock some mutexes.

> The implementation details can be found [here](https://elixir.bootlin.com/linux/v6.11.5/source/fs/file.c#L1100). I think we can do something similar in Rust (e.g. by temporarily moving the file out). Would you mind giving it a try? (Or maybe I can give it a try, since I am the one introducing the performance drop...)

Well, I can quickly assess how moving files from the file table, instead of cloning Arc, will impact performance. However, I'm concerned that the Linux approach may not apply directly to our situation.

In Linux, each thread has its own task_struct, meaning each thread has its own copy of the file table. So, checking the reference count of the file table is feasible. If the reference count is 1, no other threads can access it, making it valid to temporarily remove files from the file table. However, in our current implementation, the file table resides within the Process struct. This means that even if multiple threads within the Process can access the file table simultaneously, the reference count remains 1. If two threads perform polling concurrently, the latter thread might not see the files in the file table.

I can think of two possible solutions:

1. Move the file table from the Process to the PosixThread. This way, each thread would have its own copy of the file table, allowing the reference count check and temporary file removal to function correctly. This is the most straightforward solution.

2. Introduce a new machanism to indicate whether a struct may be shared by multiple threads or owned by a single thread. If it's shared, it must be protected with a lock and files should be cloned. Otherwise, it can be managed with a RefCell, eliminating the need to clone files. Indeed, our current `Arc<Spinlock<FileTable>>` is heavy if the file table is not shared. A simple RefCell<FileTable> would be better for performance."

Reply:
I can't see an easy way for our Rust implementation to satisfy these requirements. And by the way, Linux file table uses RCU. Can we mimic Linux's implementation without having RCU?

```c
/*
 * Lightweight file lookup - no refcnt increment if fd table isn't shared.
 *
 * You can use this instead of fget if you satisfy all of the following
 * conditions:
 * 1) You must call fput_light before exiting the syscall and returning control
 *    to userspace (i.e. you cannot remember the returned struct file * after
 *    returning to userspace).
 * 2) You must not call filp_close on the returned struct file * in between
 *    calls to fget_light and fput_light.
 * 3) You must not clone the current task in between the calls to fget_light
 *    and fput_light.
 *
 * The fput_needed flag returned by fget_light should be passed to the
 * corresponding fput_light.
 */
static unsigned long __fget_light(unsigned int fd, fmode_t mask)
{
```

Reply:
> I can think of two possible solutions:

We can check if it is a single-threaded process, right? If it is a single-threaded process, we can temporarily move the file out (and leave a placeholder in the file table so that further access will panic, indicating that our code has a bug). I guess this will just work?

Note that we need to provide an API that takes out _all_ requested files; taking them out one at a time will obviously not work if the user program requests some duplicate files.

> I can't see an easy way for our Rust implementation to satisfy these requirements.

We cannot ensure this requirements by the Rust language, but we _can_ just panic if such requirements are violated (e.g., if the file is temporarily moved out, but the file is accessed again, we'll just panic to indicate our code has bugs).

> And by the way, Linux file table uses RCU. Can we mimic Linux's implementation without having RCU?

Well, I believe RCU is not related to this topic. Please correct me if I'm wrong.

Reply:
I evaluated the performance impact of removing files from the file table and re-adding them after polling all files in `do_poll`. Unfortunately, the performance results were even worse than cloning each `Arc<FileLike>`. It appears that adding and removing files from the file table is also time-consuming. The patched code is available on [this branch](https://github.com/StevenJiang1110/asterinas/tree/remove-files-temporarily). As it is meant for quick verification, no efforts have been made to clean it up

Reply:
> I evaluated the performance impact of removing files from the file table and re-adding them after polling all files in `do_poll`. Unfortunately, the performance results were even worse than cloning each `Arc<FileLike>`. It appears that adding and removing files from the file table is also time-consuming. The patched code is available on [this branch](https://github.com/StevenJiang1110/asterinas/tree/remove-files-temporarily). As it is meant for quick verification, no efforts have been made to clean it up

It appears that you are moving out the entire `FileTableEntry`. I would guess this is not necessary? The whole `Entry` seems a bit larger to me, although it is not clear if this can cause significant overhead.

```rust
pub struct FileTableEntry {
    file: Arc<dyn FileLike>,
    flags: AtomicU8,
    subject: Subject<FdEvents>,
    owner: Option<Owner>,
}
```

Reply:
> It appears that you are moving out the entire FileTableEntry. I would guess this is not necessary? The whole Entry seems a bit larger to me, although it is not clear if this can cause significant overhead.

Well, I can give it another try to see whether only moving out `Arc<FileLike>` will be helpful.

Could you please take a look at my latest commits? I've introduced a new `ExclusiveOrShared` structure to wrap FileTable. This allows us to avoid acquiring a lock when there is only a single owner of the file table, using AtomicRefCell instead. Consequently, we can hold a reference to the file table during a poll syscall, eliminating the need for Arc::clone. As this might involve significant code changes, I've only ensured it compiles for now, and the implementation is incomplete.

Reply:
> Could you please take a look at my latest commits?

I just skimmed it. My initial feelings are as follows:
 - `ExclusiveOrShared` doubles the memory cost, because two fields `exclusive` and `shared` contain two instances of `T`.
 - [`atomic_refcell`](https://docs.rs/atomic_refcell/latest/atomic_refcell/#) contains unsafe code, so it expands our trusted base.

So I think this solution is not very ideal.

> Well, I can give it another try to see whether only moving out Arc<FileLike> will be helpful.

Before I make any sensible decisions, I'd like to know if this approach can work.

Reply:
> So I think this solution is not very ideal.

Yes, I agree. Introducing unsafe dependencies is never desirable, and the current implementation is ugly.

> Before trying to draw any meaningful conclusions, I'd like to know if this approach can work, as I think that seems to be an important data point.

I've tested it on our CI machine, and it appears to enhance performance. If with only Arc<FileLike> moved out, Asterinas will be faster than Linux (Asterinas: 2.4us, Linux: 2.7us). If you're interested, my patched code is on [this branch](https://github.com/StevenJiang1110/asterinas/tree/remove-files-temporarily), alghouth the code is incomplete now.

Do you think this is the right choice to proceed with? If so, I can clean up the code in this PR. Temporarily moving files out and adding them back is acceptable to me.

Reply:
> > And by the way, Linux file table uses RCU. Can we mimic Linux's implementation without having RCU?
> 
> Well, I believe RCU is not related to this topic. Please correct me if I'm wrong.

However, I think it is an interesting question how the above approach _interacts_ with RCU.

Basically, RCU will forbid us to get a mutable reference, which must be a hard requirement. So if we use RCU in the file table, it won't be possible for us to temporarily move the file out.

In Linux, one can simply [say](https://elixir.bootlin.com/linux/v6.11.5/source/include/linux/fdtable.h#L72):
```c
/*
 * The caller must ensure that fd table isn't shared or hold rcu or file lock
 */
static inline struct file *files_lookup_fd_raw(struct files_struct *files, unsigned int fd)
```
But in Rust, the type system cannot easily express this.

Reply:
> **EDIT:** Actually, after some more thought, I think it's still possible. This should not be a blocker for this PR.

See https://github.com/asterinas/asterinas/issues/1550.